pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/applicationconfiguration.go

```diff
@@ -576,36 +574,80 @@ func (e *GenerationUnchanged) Error() string {
 		"Please ignore this error in other logic.")
 }
 
-func applyOnceOnly() apply.ApplyOption {
-	return func(ctx context.Context, current, desired runtime.Object) error {
-		if current == nil {
+// applyOnceOnly is an ApplyOption that controls the applying mechanism for workload and trait.
+// More detail refers to ApplyOnceOnlyMode
+func applyOnceOnly(ac *v1alpha2.ApplicationConfiguration, mode core.ApplyOnceOnlyMode) apply.ApplyOption {
+	return func(_ context.Context, existing, desired runtime.Object) error {
+		if mode == core.ApplyOnceOnlyOff {
 			return nil
 		}
-		// ApplyOption only works for update/patch operation and will be ignored
-		// if the object doesn't exist before.
-		c, _ := current.(metav1.Object)
+
 		d, _ := desired.(metav1.Object)
-		if c == nil || d == nil {
-			return errors.Errorf("invalid object being applied: %q ",
+		if d == nil {
+			return errors.Errorf("cannot access metadata of object being applied: %q",
 				desired.GetObjectKind().GroupVersionKind())
 		}
-		cLabels, dLabels := c.GetLabels(), d.GetLabels()
-		if dLabels[oam.LabelOAMResourceType] == oam.ResourceTypeWorkload ||
-			dLabels[oam.LabelOAMResourceType] == oam.ResourceTypeTrait {
-			// check whether spec changes occur on the workload or trait,
-			// according to annotations and lables
-			if c.GetAnnotations()[oam.AnnotationAppGeneration] !=
-				d.GetAnnotations()[oam.AnnotationAppGeneration] {
+		dLabels := d.GetLabels()
+		if dLabels[oam.LabelOAMResourceType] != oam.ResourceTypeWorkload &&
+			dLabels[oam.LabelOAMResourceType] != oam.ResourceTypeTrait {
+			// if target resource is not a workload nor trait
+			// skip this option
+			return nil
+		}
+
+		// the resource doesn't exist (maybe not created before, or created but deleted by others)
+		if existing == nil {
+			if mode != core.ApplyOnceOnlyForce {
+				// under non-force mode, always create the resource if not exist.
 				return nil
 			}
-			if cLabels[oam.LabelAppComponentRevision] != dLabels[oam.LabelAppComponentRevision] ||
-				cLabels[oam.LabelAppComponent] != dLabels[oam.LabelAppComponent] ||
-				cLabels[oam.LabelAppName] != dLabels[oam.LabelAppName] {
-				return nil
+
+			createdBefore := false
+			for _, w := range ac.Status.Workloads {
+				if w.Reference.GetObjectKind().GroupVersionKind() == desired.GetObjectKind().GroupVersionKind() &&
+					w.Reference.Name == d.GetName() {
+					// a recorded workload matched target resource
+					createdBefore = true
+				}
+				if !createdBefore {
+					for _, t := range w.Traits {
+						if t.Reference.GetObjectKind().GroupVersionKind() == desired.GetObjectKind().GroupVersionKind() &&
+							t.Reference.Name == d.GetName() {
+							// a recorded trait matched target resource
+							createdBefore = true
+						}
+					}
+				}
+				if createdBefore {
```

Comment:
use if else in this case

Reply:
Line 617 may change `createdBefore` to true, "if else" will miss that.

Reply:
I'm wrong, it can't be else

Reply:
please add an annotation cause it's confusing.